### ISSUE TITLE
Numerical errors of intermediate hidden states

### ISSUE THREAD

**Issue by zzk1st** [HazyResearch/flash-attention#86]
Hi,

We have trained our BERT-like model using flash attention, and we want to understand the numerical errors of the intermediate hidden states with/without flash attention.

To analyze that, I've trained two head-to-head models with flash/pytorch attention, and I did inference for one batch of sequences on two models, and compare their differences on the hidden states of different layers. So the way I compare the hidden states from the two models of the same layer is, for each token, I do: L2_distance(token_embedding_model1 - token_embedding_model2) / L2_norm_token_embedding_model1, and then I average all the tokens of this sequence, making it a token-wise RTOL.

What I found is the RTOL of the hiddens states in the layers are pretty high (over 100%), but the RTOL of the final logits are only 5%, which makes the output of the model with flash attention not so much difference from that of pytorch. My question is have you guys done analysis related to this topic before? What could we possibly do to reduce the RTOL of the hidden states? Thanks in advance!

**Comment by tridao** [comment]
Even the same model code trained with 2 different seeds would probably produce very different intermediate representations (e.g, you can double the weight matrix of one layer and halve the weight of another layer to produce the same output).

Just so I understand your comparison procedure: if you train 2 different models using 2 different ways of implementing attention in Pytorch (just swapping the order of operations), would the intermediate representations be the same?
```
scores = torch.einsum('bshd,bthd->bhst', q / math.sqrt(d), k)
```
vs 
```
scores = torch.einsum('bshd,bthd->bhst', q, k / math.sqrt(d))
```
Similarly, if you train two models, both with Pytorch code, but with different seeds, would the intermediate representations be the same?

**Comment by zzk1st** [comment]
Thanks for the reply! The two models we train has exactly the same hyperparams/seeds/data etc, the only difference is one is using flash attention, the other is standard pytorch attention. Correct me if I'm wrong, my understanding is flash attention and pytorch attention they are mathematically equivalent, so they should be producing the same intermediate results?

**Comment by tridao** [comment]
I don't expect the same intermediate representations if you train the same models with two different implementations.

As an example, if you replace your Pytorch attention code with a mathematically equivalent implementation below, once you train it, I don't expect the intermediate representations to be the same.
Could you measure the error for this implementation?
```
new_q = q / 0.6
new_k = k * 0.6
return attention(new_q, new_k, v)
```

**Comment by zzk1st** [comment]
I see, thanks for the clarification!
